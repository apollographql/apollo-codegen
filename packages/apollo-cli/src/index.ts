// run the CLI
export { run } from "@oclif/command";
