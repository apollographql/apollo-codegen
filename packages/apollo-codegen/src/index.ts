import 'core-js/fn/object/values';
import 'core-js/fn/object/entries';

export { default as downloadSchema } from './downloadSchema';
export { default as introspectSchema } from './introspectSchema';
export { default as printSchema } from './printSchema';
export { default as generate } from './generate';
