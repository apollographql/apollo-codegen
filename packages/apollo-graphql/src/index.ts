export {
  defaultEngineReportingSignature,
  defaultOperationRegistrySignature,
  operationRegistrySignature,
  operationHash
} from "./operationId";
export * from "./schema";
export {
  printWithReducedWhitespace,
  hideStringAndNumericLiterals,
  hideLiterals
} from "./transforms";
