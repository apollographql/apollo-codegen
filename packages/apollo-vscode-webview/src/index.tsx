import * as React from "react";
import * as ReactDOM from "react-dom";
import App from "./App";

declare function acquireVsCodeApi(): {
  postMessage(message: any): void;
};

export const vscode = acquireVsCodeApi();

ReactDOM.render(<App />, document.getElementById("root") as HTMLElement);
