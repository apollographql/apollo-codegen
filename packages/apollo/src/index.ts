// run the CLI
export { run } from "@oclif/command";

// export utils for usage
export * from "./printer/ast";

export { printFromSchemas, printChanges } from "./printer/print";
export { diffSchemas } from "./printer/diff";
