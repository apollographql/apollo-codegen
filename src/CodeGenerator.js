import {
  join,
  wrap,
} from './utilities/printing';

export default class CodeGenerator {
  constructor() {
    this.scopeStack = [];

    this.indentWidth = 2;
    this.indentLevel = 0;

    this.output = '';
  }

  pushScope(scope) {
    this.scopeStack.push(scope);
  }

  popScope() {
    return this.scopeStack.pop();
  }

  print(maybeString) {
    if (maybeString) {
      this.output += maybeString;
    }
  }

  printNewline() {
    if (this.output) {
      this.print('\n');
      this.startOfIndentLevel = false;
    }
  }

  printNewlineIfNeeded() {
    if (!this.startOfIndentLevelLevel) {
      this.printNewline();
    }
  }

  printOnNewline(maybeString) {
    if (maybeString) {
      this.printNewline();
      this.printIndent();
      this.print(maybeString);
    }
  }

  printIndent() {
    const indentation = ' '.repeat(this.indentLevel * this.indentWidth);
    this.output += indentation;
  }

  withIndent(closure) {
    if (!closure) return;

    this.indentLevel++;
    this.startOfIndentLevel = true;
    closure();
    this.indentLevel--;
  }

  withinBlock(closure) {
    this.print(' {');
    this.withIndent(closure);
    this.printOnNewline('}');
  }
}
