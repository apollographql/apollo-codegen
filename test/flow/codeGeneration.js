import {
  parse,
  isType,
  GraphQLID,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull
} from 'graphql';

import {
  generateSource
} from '../../src/flow/codeGeneration';

import { loadSchema } from '../../src/loading';
const swapiSchema = loadSchema(require.resolve('../starwars/schema.json'));
const miscSchema = loadSchema(require.resolve('../misc/schema.json'));

import CodeGenerator from '../../src/utilities/CodeGenerator';

import { compileToIR } from '../../src/compilation';

function setup(schema) {
  const context = {
    schema: schema,
    operations: {},
    fragments: {},
    typesUsed: {}
  }

  const generator = new CodeGenerator(context);

  const compileFromSource = (source, addTypename = false) => {
    const document = parse(source);
    const context = compileToIR(schema, document);
    context.addTypename = addTypename;
    generator.context = context;
    return context;
  };

  const addFragment = (fragment) => {
    generator.context.fragments[fragment.fragmentName] = fragment;
  };

  return { generator, compileFromSource, addFragment };
}

describe('Flow code generation', function() {
  describe('#generateSource()', function() {
    describe('__typename', function() {
      test('in an object', function() {
        const { compileFromSource } = setup(swapiSchema);
        const context = compileFromSource(`
          query HeroName {
            hero {
              __typename
              name
            }
          }
        `);

        const source = generateSource(context);
        expect(source).toMatchSnapshot();
      });

      // TODO: Enable after fixing operation
      test.skip('in an operation', function() {
        const { compileFromSource } = setup(swapiSchema);
        const context = compileFromSource(`
          query Hero {
            ...Hero
          }

          fragment Hero on Query {
            hero {
              __typename
              name
            }
          }
        `);

        const source = generateSource(context);
        expect(source).toMatchSnapshot();
      });

      test('single fragment spread', () => {
        const { compileFromSource } = setup(swapiSchema);
        const context = compileFromSource(`
          query HeroName {
            hero {
              ...humanFriends
            }
          }

          fragment humanFriends on Character {
            friends {
              __typename
              name
            }
          }
        `);

        const source = generateSource(context);
        expect(source).toMatchSnapshot();
      });

      test('in fragment spreads, allows for disjoint union via __typename string literals', function() {
        const { compileFromSource } = setup(swapiSchema);
        const context = compileFromSource(`
          query HeroName {
            hero {
              __typename
              ...humanHero
              ...droidHero
            }
          }

          fragment droidHero on Droid {
            primaryFunction
          }

          fragment humanHero on Human {
            homePlanet
          }
        `);

        const source = generateSource(context);
        expect(source).toMatchSnapshot();
      });

      test('in inline fragments, allows for disjoint union via __typename string literals', function() {
        const { compileFromSource } = setup(swapiSchema);
        const context = compileFromSource(`
          query HeroName {
            hero {
              __typename

              ... on Droid {
                friends {
                  name
                }
              }
              ... on Human {
                homePlanet
              }
            }
          }
        `);

        const source = generateSource(context);
        expect(source).toMatchSnapshot();
      });
    });

    test(`should generate simple query operations`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroName {
          hero {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate simple query operations including input variables`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroName($episode: Episode) {
          hero(episode: $episode) {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate simple nested query operations including input variables`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              name
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate fragmented query operations`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndFriendsNames {
          hero {
            name
            ...heroFriends
          }
        }

        fragment heroFriends on Character {
          friends {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should handle multi-fragmented query operations`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndFriendsNames {
          hero {
            name
            ...heroFriends
            ...heroAppears
          }
        }

        fragment heroFriends on Character {
          friends {
            name
          }
        }

        fragment heroAppears on Character {
          appearsIn
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate query operations with inline fragments`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndDetails {
          hero {
            name
            ...HeroDetails
          }
        }

        fragment HeroDetails on Character {
          ... on Droid {
            primaryFunction
          }
          ... on Human {
            height
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate mutation operations with complex input types`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        mutation ReviewMovie($episode: Episode, $review: ReviewInput) {
          createReview(episode: $episode, review: $review) {
            stars
            commentary
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate correct typedefs with a single custom fragment`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        fragment Friend on Character {
          name
        }

        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              ...Friend
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate correct typedefs with a multiple custom fragments`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        fragment Friend on Character {
          name
        }

        fragment Person on Character {
          name
        }

        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              ...Friend
              ...Person
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should handle complex fragments with type aliases`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndFriendsNames {
          hero(episode: NEWHOPE) {
            name
            ...Something
          }
          empireHero: hero(episode: EMPIRE) {
            name
            ...Something
          }
        }

        fragment Something on Character {
          ... on Human {
            friends {
              ... on Human {
                homePlanet
              }

              ... on Droid {
                primaryFunction
              }
            }
          }

          ... on Droid {
            appearsIn
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should annotate custom scalars as string`, function() {
      const { compileFromSource } = setup(miscSchema);
      const context = compileFromSource(`
        query CustomScalar {
          misc {
            date
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test('should correctly handle fragments on interfaces', function() {
      const {compileFromSource} = setup(swapiSchema);
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode){
          hero(episode: $episode) {
            name
            ... on Human {
              homePlanet
            }

            ... on Droid {
              primaryFunction
            }
          }
        }
      `
      );

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test('should correctly handle fragment spreads on interfaces', function() {
      const {compileFromSource} = setup(swapiSchema);
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode){
          hero(episode: $episode) {
            name
            ...humanFragment
            ...droidFragment
          }
        }

        fragment humanFragment on Human {
          homePlanet
        }

        fragment droidFragment on Droid {
          primaryFunction
        }
      `
      );

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test('should correctly handle nested fragments on interfaces', function() {
      const {compileFromSource} = setup(swapiSchema);
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode){
          hero(episode: $episode) {
            name
            friendsConnection {
              friends {
                ...CharacterFragment
              }
            }
          }
        }

        fragment CharacterFragment on Character {
          name

          ... on Human {
            homePlanet
          }

          ... on Droid {
            primaryFunction
          }
        }
      `
      );

      const source = generateSource(context);

      expect(source).toMatchSnapshot();
    });

    test('should correctly add typename to nested fragments on interfaces if addTypename is true', function() {
      const {compileFromSource} = setup(swapiSchema);
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode){
          hero(episode: $episode) {
            name
            friendsConnection {
              friends {
                ...CharacterFragment
              }
            }
          }
        }

        fragment CharacterFragment on Character {
          name

          ... on Human {
            homePlanet
          }

          ... on Droid {
            primaryFunction
          }
        }
      `
      );

      const source = generateSource(context, { addTypename: true });

      expect(source).toMatchSnapshot();
    });

    test('should correctly handle doubly nested fragments on interfaces', function() {
      const {compileFromSource} = setup(swapiSchema);
      const context = compileFromSource(
        `
        query HeroQuery($episode: Episode) {
          hero(episode: $episode) {
            name
            friendsConnection {
              friends {
                ...CharacterFragment
              }
            }
          }
        }

        fragment CharacterFragment on Character {
          name

          ... on Human {
            homePlanet
            friends {
              ...OtherCharacterFragment
            }
          }

          ... on Droid {
            primaryFunction
          }
        }

        fragment OtherCharacterFragment on Character {
          ... on Human {
            height
          }

          ... on Droid {
            appearsIn
          }
        }
      `
      );

      const source = generateSource(context);

      expect(source).toMatchSnapshot();
    });
  });
});
