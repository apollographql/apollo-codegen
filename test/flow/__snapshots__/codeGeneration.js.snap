// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow code generation #generateSource() __typename in an object 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroNameQuery = {|
  hero: ? {|
    __typename: string,
    // The name of the character
    name: string,
  |},
|};"
`;

exports[`Flow code generation #generateSource() __typename in an operation 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroQuery = {|
|};

export type HeroFragment = {|
  hero: ? {|
    __typename: string,
    // The name of the character
    name: string,
  |},
|};"
`;

exports[`Flow code generation #generateSource() __typename in fragment spreads, allows for disjoint union via __typename string literals 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroNameQuery = {|
  hero: ? {|
    ...HumanHeroFragment,
    ...DroidHeroFragment,
    __typename: string,
  |},
|};

export type DroidHeroFragment = {|
  // This droid's primary function
  primaryFunction: ?string,
|};

export type HumanHeroFragment = {|
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
|};"
`;

exports[`Flow code generation #generateSource() __typename in inline fragments, allows for disjoint union via __typename string literals 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroNameQuery = {|
  hero: ?HeroFromHeroName,
|};

export type HeroFromHeroNameOnDroid = {|
  __typename: string,
  // This droid's friends, or an empty list if they have none
  friends: ?Array< {|
    // The name of the character
    name: string,
  |} >,
|};

export type HeroFromHeroNameOnHuman = {|
  __typename: string,
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
|};

export type HeroFromHeroName =
  | HeroFromHeroNameOnDroid
  | HeroFromHeroNameOnHuman;"
`;

exports[`Flow code generation #generateSource() __typename single fragment spread 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroNameQuery = {|
  hero: ?HumanFriendsFragment,
|};

export type HumanFriendsFragment = {|
  // The friends of the character, or an empty list if they have none
  friends: ?Array< {|
    __typename: string,
    // The name of the character
    name: string,
  |} >,
|};"
`;

exports[`Flow code generation #generateSource() should annotate custom scalars as string 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type CustomScalarQuery = {|
  misc: ? {|
    date: ?any,
  |},
|};"
`;

exports[`Flow code generation #generateSource() should correctly add typename to nested fragments on interfaces if addTypename is true 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroQueryQueryVariables = {|
  episode: ?Episode,
|};

export type HeroQueryQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
    // The friends of the character exposed as a connection with edges
    friendsConnection: {|
      // A list of the friends, as a convenience when edges are not needed.
      friends: ?Array<CharacterFragmentFragment>,
    |},
  |},
|};

export type CharacterFragmentFragmentOnHuman = {|
  // What this human calls themselves
  name: string,
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
|};

export type CharacterFragmentFragmentOnDroid = {|
  // What others call this droid
  name: string,
  // This droid's primary function
  primaryFunction: ?string,
|};

export type CharacterFragmentFragment =
  | CharacterFragmentFragmentOnHuman
  | CharacterFragmentFragmentOnDroid;"
`;

exports[`Flow code generation #generateSource() should correctly handle doubly nested fragments on interfaces 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroQueryQueryVariables = {|
  episode: ?Episode,
|};

export type HeroQueryQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
    // The friends of the character exposed as a connection with edges
    friendsConnection: {|
      // A list of the friends, as a convenience when edges are not needed.
      friends: ?Array<CharacterFragmentFragment>,
    |},
  |},
|};

export type CharacterFragmentFragmentOnHuman = {|
  // What this human calls themselves
  name: string,
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
  // This human's friends, or an empty list if they have none
  friends: ?Array<OtherCharacterFragmentFragment>,
|};

export type CharacterFragmentFragmentOnDroid = {|
  // What others call this droid
  name: string,
  // This droid's primary function
  primaryFunction: ?string,
|};

export type CharacterFragmentFragment =
  | CharacterFragmentFragmentOnHuman
  | CharacterFragmentFragmentOnDroid;

export type OtherCharacterFragmentFragmentOnHuman = {|
  // Height in the preferred unit, default is meters
  height: ?number,
|};

export type OtherCharacterFragmentFragmentOnDroid = {|
  // The movies this droid appears in
  appearsIn: Array< ?Episode >,
|};

export type OtherCharacterFragmentFragment =
  | OtherCharacterFragmentFragmentOnHuman
  | OtherCharacterFragmentFragmentOnDroid;"
`;

exports[`Flow code generation #generateSource() should correctly handle fragment spreads on interfaces 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroQueryQueryVariables = {|
  episode: ?Episode,
|};

export type HeroQueryQuery = {|
  hero: ? {|
    ...HumanFragmentFragment,
    ...DroidFragmentFragment,
    // The name of the character
    name: string,
  |},
|};

export type HumanFragmentFragment = {|
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
|};

export type DroidFragmentFragment = {|
  // This droid's primary function
  primaryFunction: ?string,
|};"
`;

exports[`Flow code generation #generateSource() should correctly handle fragments on interfaces 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroQueryQueryVariables = {|
  episode: ?Episode,
|};

export type HeroQueryQuery = {|
  hero: ?HeroFromHeroQuery,
|};

export type HeroFromHeroQueryOnHuman = {|
  // What this human calls themselves
  name: string,
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
|};

export type HeroFromHeroQueryOnDroid = {|
  // What others call this droid
  name: string,
  // This droid's primary function
  primaryFunction: ?string,
|};

export type HeroFromHeroQuery =
  | HeroFromHeroQueryOnHuman
  | HeroFromHeroQueryOnDroid;"
`;

exports[`Flow code generation #generateSource() should correctly handle nested fragments on interfaces 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroQueryQueryVariables = {|
  episode: ?Episode,
|};

export type HeroQueryQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
    // The friends of the character exposed as a connection with edges
    friendsConnection: {|
      // A list of the friends, as a convenience when edges are not needed.
      friends: ?Array<CharacterFragmentFragment>,
    |},
  |},
|};

export type CharacterFragmentFragmentOnHuman = {|
  // What this human calls themselves
  name: string,
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
|};

export type CharacterFragmentFragmentOnDroid = {|
  // What others call this droid
  name: string,
  // This droid's primary function
  primaryFunction: ?string,
|};

export type CharacterFragmentFragment =
  | CharacterFragmentFragmentOnHuman
  | CharacterFragmentFragmentOnDroid;"
`;

exports[`Flow code generation #generateSource() should generate correct typedefs with a multiple custom fragments 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroAndFriendsNamesQueryVariables = {|
  episode: ?Episode,
|};

export type HeroAndFriendsNamesQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
    // The friends of the character, or an empty list if they have none
    friends: ?Array<{|
      ...FriendFragment,
      ...PersonFragment,
    |}>,
  |},
|};

export type FriendFragment = {|
  // The name of the character
  name: string,
|};

export type PersonFragment = {|
  // The name of the character
  name: string,
|};"
`;

exports[`Flow code generation #generateSource() should generate correct typedefs with a single custom fragment 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroAndFriendsNamesQueryVariables = {|
  episode: ?Episode,
|};

export type HeroAndFriendsNamesQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
    // The friends of the character, or an empty list if they have none
    friends: ?Array<FriendFragment>,
  |},
|};

export type FriendFragment = {|
  // The name of the character
  name: string,
|};"
`;

exports[`Flow code generation #generateSource() should generate fragmented query operations 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroAndFriendsNamesQuery = {|
  hero: ? {|
    ...HeroFriendsFragment,
    // The name of the character
    name: string,
  |},
|};

export type HeroFriendsFragment = {|
  // The friends of the character, or an empty list if they have none
  friends: ?Array< {|
    // The name of the character
    name: string,
  |} >,
|};"
`;

exports[`Flow code generation #generateSource() should generate mutation operations with complex input types 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type ReviewInput = {|
  // 0-5 stars
  stars: number,
  // Comment about the movie, optional
  commentary: ?string,
  // Favorite color, optional
  favorite_color: ?ColorInput,
|};

export type ColorInput = {|
  red: number,
  green: number,
  blue: number,
|};

export type ReviewMovieMutationVariables = {|
  episode: ?Episode,
  review: ?ReviewInput,
|};

export type ReviewMovieMutation = {|
  createReview: ? {|
    // The number of stars this review gave, 1-5
    stars: number,
    // Comment about the movie
    commentary: ?string,
  |},
|};"
`;

exports[`Flow code generation #generateSource() should generate query operations with inline fragments 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroAndDetailsQuery = {|
  hero: ? {|
    ...HeroDetailsFragment,
    // The name of the character
    name: string,
  |},
|};

export type HeroDetailsFragmentOnDroid = {|
  // This droid's primary function
  primaryFunction: ?string,
|};

export type HeroDetailsFragmentOnHuman = {|
  // Height in the preferred unit, default is meters
  height: ?number,
|};

export type HeroDetailsFragment =
  | HeroDetailsFragmentOnDroid
  | HeroDetailsFragmentOnHuman;"
`;

exports[`Flow code generation #generateSource() should generate simple nested query operations including input variables 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroAndFriendsNamesQueryVariables = {|
  episode: ?Episode,
|};

export type HeroAndFriendsNamesQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
    // The friends of the character, or an empty list if they have none
    friends: ?Array< {|
      // The name of the character
      name: string,
    |} >,
  |},
|};"
`;

exports[`Flow code generation #generateSource() should generate simple query operations 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

export type HeroNameQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
  |},
|};"
`;

exports[`Flow code generation #generateSource() should generate simple query operations including input variables 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroNameQueryVariables = {|
  episode: ?Episode,
|};

export type HeroNameQuery = {|
  hero: ? {|
    // The name of the character
    name: string,
  |},
|};"
`;

exports[`Flow code generation #generateSource() should handle complex fragments with type aliases 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroAndFriendsNamesQuery = {|
  hero: ? {|
    ...SomethingFragment,
    // The name of the character
    name: string,
  |},
  empireHero: ? {|
    ...SomethingFragment,
    // The name of the character
    name: string,
  |},
|};

export type SomethingFragmentOnHuman = {|
  // This human's friends, or an empty list if they have none
  friends: ?Array< ?FriendFragment >,
|};

export type FriendFragmentOnHuman = {|
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
|};

export type FriendFragmentOnDroid = {|
  // This droid's primary function
  primaryFunction: ?string,
|};

export type FriendFragment =
  | FriendFragmentOnHuman
  | FriendFragmentOnDroid;

export type SomethingFragmentOnDroid = {|
  // The movies this droid appears in
  appearsIn: Array< ?Episode >,
|};

export type SomethingFragment =
  | SomethingFragmentOnHuman
  | SomethingFragmentOnDroid;"
`;

exports[`Flow code generation #generateSource() should handle multi-fragmented query operations 1`] = `
"/* @flow */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroAndFriendsNamesQuery = {|
  hero: ? {|
    ...HeroFriendsFragment,
    ...HeroAppearsFragment,
    // The name of the character
    name: string,
  |},
|};

export type HeroFriendsFragment = {|
  // The friends of the character, or an empty list if they have none
  friends: ?Array< {|
    // The name of the character
    name: string,
  |} >,
|};

export type HeroAppearsFragment = {|
  // The movies this character appears in
  appearsIn: Array< ?Episode >,
|};"
`;
